src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,530 @@
+                        }
+                    }
+
+                    persist_state_per_vnode(
+                        barrier.epoch,
+                        &mut self.state_table,
+                        false,
+                        &mut backfill_state,
+                        &mut committed_progress,
+                        &mut temporary_state,
+                    ).await?;
+
+                    self.progress.finish(barrier.epoch.curr);
+                    yield msg;
+                    break;
+                }
+                yield msg;
+            }
+        }
+
+        // After progress finished + state persisted,
+        // we can forward messages directly to the downstream,
+        // as backfill is finished.
+        #[for_await]
+        for msg in upstream {
+            if let Some(msg) = mapping_message(msg?, &self.output_indices) {
+                if let Message::Barrier(barrier) = &msg {
+                    self.state_table.commit_no_data_expected(barrier.epoch);
+                }
+                yield msg;
+            }
+        }
+    }
+
+    /// Read snapshot per vnode.
+    /// These streams should be sorted in storage layer.
+    /// 1. Get row iterator / vnode.
+    /// 2. Merge it with `select_all`.
+    /// 3. Change it into a chunk iterator with `iter_chunks`.
+    /// This means it should fetch a row from each iterator to form a chunk.
+    ///
+    /// NOTE(kwannoel): We interleave at chunk per vnode level rather than rows.
+    /// This is so that we can compute `current_pos` once per chunk, since they correspond to 1
+    /// vnode.
+    ///
+    /// TODO(kwannoel): Support partially complete snapshot reads.
+    /// That will require the following changes:
+    /// Instead of returning stream chunk and vnode, we need to dispatch 3 diff messages:
+    /// 1. COMPLETE_VNODE(vnode): Current iterator is complete, in that case we need to handle it
+    ///    in arrangement backfill. We should not buffer updates for this vnode, and forward
+    ///    all messages.
+    /// 2. MESSAGE(CHUNK): Current iterator is not complete, in that case we
+    ///    need to buffer updates for this vnode.
+    /// 3. FINISHED: All iterators finished.
+    ///
+    /// For now we only support the case where all iterators are complete.
+    #[try_stream(ok = Option<(VirtualNode, StreamChunk)>, error = StreamExecutorError)]
+    async fn snapshot_read_per_vnode(
+        schema: Arc<Schema>,
+        upstream_table: &ReplicatedStateTable<S>,
+        backfill_state: BackfillState,
+    ) {
+        let mut streams = Vec::with_capacity(upstream_table.vnodes().len());
+        for vnode in upstream_table.vnodes().iter_vnodes() {
+            let backfill_progress = backfill_state.get_progress(&vnode)?;
+            let current_pos = match backfill_progress {
+                BackfillProgressPerVnode::Completed => {
+                    continue;
```

Comment:
We no longer `Yield None`, since these are separate iterators, and if 1 of them yield none, it will stop the backfill entirely.

Instead we wait until there are no more chunks, and yield `None` below.

Reply:
https://github.com/risingwavelabs/risingwave/pull/10266/files#r1251678447